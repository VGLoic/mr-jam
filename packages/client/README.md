## Available Scripts

In the project directory, you can run:

### `yarn install`

Installs the dependencies for the client application.

### `yarn start:client`

Runs the app in the development mode.<br />
Open [http://localhost:3000](http://localhost:3000) to view it in the browser.

The page will reload if you make edits.<br />
You will also see any lint errors in the console.

You will need the lambda functions running locally in order to use the application. See next section.

### `yarn test`

Launches the test runner for the client application in the interactive watch mode.<br />
See the section about [running tests](https://facebook.github.io/create-react-app/docs/running-tests) for more information.

### `yarn test:cover`

Launches the test runner and generates coverage for the client.

### `yarn build`

Builds the client for production to the `packages/client/build` folder.<br />
It correctly bundles React in production mode and optimizes the build for the best performance.

The build is minified and the filenames include the hashes.<br />
Your app is ready to be deployed!

See the section about [deployment](https://facebook.github.io/create-react-app/docs/deployment) for more information.


