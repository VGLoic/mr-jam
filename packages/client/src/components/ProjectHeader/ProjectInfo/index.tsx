import React from "react";
// UI Components
import { CircularProgress, Typography, Button, Box } from "@material-ui/core";
// Components
import CreateProjectDialog from "./CreateProjectDialog";
import ManageProjectDialog from "./ManageProjectDialog";
// Hooks
import useDialog from "hooks/useDialog";
// Types
import { User } from "types/user";
import { UseEthereumProject } from "pages/ProjectPage/controllers/useEthereumProject";

interface ProjectInfoProps {
  ethereumProject: UseEthereumProject;
  projectUsers: User[];
  projectId: string;
}
const ProjectMembership = ({
  ethereumProject,
  projectUsers,
  projectId,
}: ProjectInfoProps) => {
  const { open, closeDialog, openDialog } = useDialog();

  const {
    unable,
    loading,
    error,
    projectAddress,
    refetchProjectAddress,
    isMember,
    isAdmin,
  } = ethereumProject;

  if (unable) return null;

  if (loading) {
    return (
      <CircularProgress
        data-testid="project-address-loading"
        aria-label="Searching for existing project"
      />
    );
  }

  if (error) {
    return (
      <Typography
        data-testid="project-address-error"
        aria-label="There has been an error... :("
        color="textPrimary"
      >
        There has been an error... :(
      </Typography>
    );
  }

  if (!projectAddress) {
    return (
      <>
        <Button
          onClick={openDialog}
          variant="contained"
          aria-label="Open project creation dialog"
          data-testid="project-creation-dialog-button"
        >
          Click here to setup on Ethereum
        </Button>
        <CreateProjectDialog
          open={open}
          closeDialog={closeDialog}
          updateProjectAddress={refetchProjectAddress}
          projectUsers={projectUsers}
          projectId={projectId}
        />
      </>
    );
  }

  if (!isMember) {
    return null;
  }

  return (
    <Box display="flex" alignItems="center">
      <Typography
        data-testid="member"
        aria-label="You are a member"
        color="textPrimary"
      >
        Youpi
      </Typography>
      {isAdmin && (
        <>
          <Button
            onClick={openDialog}
            variant="contained"
            aria-label="Open project management dialog"
            data-testid="project-management-dialog-button"
          >
            Manage Project
          </Button>
          <ManageProjectDialog
            open={open}
            closeDialog={closeDialog}
            projectAddress={projectAddress}
            projectUsers={projectUsers}
          />
        </>
      )}
    </Box>
  );
};

export default ProjectMembership;
