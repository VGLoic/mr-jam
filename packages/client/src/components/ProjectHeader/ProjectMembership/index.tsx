import React from "react";
// UI Components
import { CircularProgress, Typography, Button } from "@material-ui/core";
// Components
import CreateProjectDialog from "./CreateProjectDialog";
// Hooks
import useProjectMembership from "./controllers/useProjectMembership";
import useDialog from "hooks/useDialog";
// Types
import { User } from "types/user";

interface ProjectMembershipProps {
  projectName: string;
  projectUsers: User[];
}
const ProjectMembership = ({
  projectName,
  projectUsers,
}: ProjectMembershipProps) => {
  const {
    refetch,
    unable,
    loading,
    error,
    projectAddress,
  } = useProjectMembership(projectName);

  const { open, closeDialog, openDialog } = useDialog();

  if (unable) return null;

  if (loading) {
    return (
      <CircularProgress
        data-testid="project-address-loading"
        aria-label="Searching for existing project"
      />
    );
  }

  if (error) {
    return (
      <Typography
        data-testid="project-address-error"
        aria-label="There has been an error... :("
        color="textPrimary"
      >
        There has been an error... :(
      </Typography>
    );
  }

  if (!projectAddress) {
    return (
      <>
        <Button onClick={openDialog} variant="contained">
          Click here to setup on Ethereum
        </Button>
        <CreateProjectDialog
          open={open}
          closeDialog={closeDialog}
          updateProjectAddress={refetch}
          projectName={projectName}
          projectUsers={projectUsers}
        />
      </>
    );
  }

  return <Typography color="textPrimary">Yeay</Typography>;
};

export default ProjectMembership;
