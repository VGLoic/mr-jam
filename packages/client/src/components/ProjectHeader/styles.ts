import { makeStyles, Theme } from "@material-ui/core";

const styles = ({ spacing, shadows }: Theme): Styles => ({
  avatar: {
    height: "25px",
    width: "25px",
    boxShadow: shadows[5]
  },
  avatarWrapper: {
    display: "flex",
  },
  seeMore: {
    "&:hover": {
      cursor: "pointer",
    },
  },
  projectTitle: {
    marginRight: `${spacing(2)}px`,
  },
  fullPathLabel: {
    marginRight: `${spacing(1)}px`,
  },
});

export interface Styles {
  avatar: any;
  avatarWrapper: any;
  projectTitle: any;
  fullPathLabel: any;
  seeMore: any;
}

export const useStyles = makeStyles(styles);
