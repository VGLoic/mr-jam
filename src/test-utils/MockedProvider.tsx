import React from "react";
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import DateFnsUtils from '@date-io/date-fns';
import CssBaseline from "@material-ui/core/CssBaseline";
import { BrowserRouter as Router } from "react-router-dom";
import { ThemeProvider } from "contexts/theme";


export interface MockedProviderProps {
    children?: React.ReactElement;
}
const MockedProvider = ({ children }: MockedProviderProps) => {
    return (
        <Router>
            <MuiPickersUtilsProvider utils={DateFnsUtils}>
                <CssBaseline />
                <ThemeProvider>
                    {children}
                </ThemeProvider>
            </MuiPickersUtilsProvider>
        </Router>
    )
}

export default MockedProvider;