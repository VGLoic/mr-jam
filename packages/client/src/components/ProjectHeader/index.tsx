import React from "react";
import classnames from "classnames";
import { useQuery } from "@apollo/client";
// UI Components
import {
  Typography,
  Grid,
  Avatar,
  Tooltip,
  Dialog,
  DialogTitle,
  List,
  ListItem,
  ListItemAvatar,
  ListItemText,
  CircularProgress
} from "@material-ui/core";
import AvatarGroup from "@material-ui/lab/AvatarGroup";
// Hooks
import { useDialog } from "hooks/useDialog";
// Query
import {
  PROJECT_OVERVIEW
} from "./controllers/projectOverview.query";
// Types
import {
  ProjectOverviewData,
  ProjectOverviewInput
} from "./controllers/projectOverview.types";
// Styles
import { useStyles, Styles } from "./styles";

interface ProjectHeaderProps {
  projectId: string;
}
const ProjectHeader = ({ projectId }: ProjectHeaderProps) => {
  const { data, loading, error } = useQuery<ProjectOverviewData, ProjectOverviewInput>(
    PROJECT_OVERVIEW,
    {
      variables: { projectId }
    }
  );
  const { open, toggleDialog } = useDialog();
  const classes: Styles = useStyles();

  if (loading) {
    return (
      <Grid item>
        <CircularProgress
          data-testid="project-overview-loading"
          aria-label="Project overview is loading"
        />
      </Grid>
    );
  }

  if (error || !data?.project) {
    return (
      <Grid item>
        <Typography
          variant="h6"
          component="h6"
          color="textPrimary"
          data-testid="project-overview-error"
          aria-label={`Project with id ${projectId} has not been found`}
        >
          This project has not been found. Please select another one.
        </Typography>
      </Grid>
    );
  }

  const project = data.project;

  const tooMuchUsers: boolean = project.users.length > 5;

  return (
    <>
      <Grid item container xs={12} direction="row" justify="space-between">
        <Grid item container sm={8}>
          <Typography
            variant="h6"
            component="h6"
            className={classes.projectTitle}
            color="textPrimary"
            data-testid="project-name"
            aria-label={`Project name: ${project.name.toUpperCase()}`}
          >
            {project.name.toUpperCase()}
          </Typography>
          <AvatarGroup>
            {tooMuchUsers ? (
              <div className={classes.avatarWrapper}>
                {project.users.slice(0, 5).map(user => (
                  <Tooltip title={user.name} key={user.id}>
                    <Avatar
                      className={classes.avatar}
                      alt={user.name}
                      src={user.avatarUrl}
                      data-testid="user-avatar"
                    />
                  </Tooltip>
                ))}
                <Avatar
                  className={classnames(classes.avatar, classes.seeMore)}
                  onClick={toggleDialog}
                  data-testid="avatar-open-dialog"
                  aria-label="Open project users dialog"
                >
                  <Typography
                    variant="subtitle2"
                    color="textPrimary"
                  >
                    +{project.users.length - 5}
                  </Typography>
                </Avatar>
              </div>
            ) : (
              project.users.map(user => (
                <Tooltip title={user.name} key={user.id}>
                  <Avatar
                    className={classes.avatar}
                    alt={user.name}
                    src={user.avatarUrl}
                    data-testid="user-avatar"
                  />
                </Tooltip>
              ))
            )}
          </AvatarGroup>
        </Grid>
        <Grid item container sm={4} justify="flex-end">
          <Typography
            variant="body1"
            className={classes.fullPathLabel}
            color="textPrimary"
          >
            Full path:
          </Typography>
          <Typography variant="body1" color="textSecondary" data-testid="full-path" aria-label={`Full path: ${project.pathWithNamespace}`}>
            {project.pathWithNamespace}
          </Typography>
        </Grid>
      </Grid>
      <Dialog open={open} onClose={toggleDialog}>
        <DialogTitle>Project users</DialogTitle>
        <List>
          {project.users.map(user => (
            <ListItem key={user.id}>
              <ListItemAvatar>
                <Avatar
                  className={classes.avatar}
                  alt={user.name}
                  src={user.avatarUrl}
                  key={user.id}
                  data-testid="avatar-dialog"
                />
              </ListItemAvatar>
              <ListItemText aria-label={`name: ${user.name}`}>{user.name}</ListItemText>
            </ListItem>
          ))}
        </List>
      </Dialog>
    </>
  );
};

export default ProjectHeader;
