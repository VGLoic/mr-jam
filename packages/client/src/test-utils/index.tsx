import React from "react";
import { render } from '@testing-library/react'
import MockedProvider from "./MockedProvider";

const customRender = (ui: React.ReactElement, options?: any) => {
  const RenderedMockedProvider = (props: any) => <MockedProvider {...props} ethersContext={options?.ethersContext} />;
  return render(ui, { wrapper: RenderedMockedProvider, ...options });
}

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }