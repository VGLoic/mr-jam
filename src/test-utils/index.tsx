import React from "react";
import { render } from '@testing-library/react'
import MockedProvider from "./MockedProvider";

const customRender = (ui: React.ReactElement, options?: any) =>
  render(ui, { wrapper: MockedProvider, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }