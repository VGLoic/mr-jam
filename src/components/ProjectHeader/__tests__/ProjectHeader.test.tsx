import React from "react";
import { render, waitFor, screen, fireEvent } from "test-utils";
import { MockedProvider } from '@apollo/client/testing';

import ProjectHeader from "../";
import { PROJECT_OVERVIEW } from "../controllers/projectOverview.query";
import { ProjectOverviewData } from "../controllers/projectOverview.types";

describe("<ProjectHeader />", () => {
    it("should render with error", async () => {
        const projectOverviewMock = {
            request: {
                query: PROJECT_OVERVIEW,
                variables: { projectId: "projectIdTest" }
            },
            error: new Error('aw shucks')
        };
        const { asFragment } = render(
            <MockedProvider mocks={[projectOverviewMock]} addTypename={false}>
                <ProjectHeader projectId="projectIdTest" />
            </MockedProvider>
        );

        await waitFor(() => screen.getByLabelText("Project with id projectIdTest has not been found"));

        expect(asFragment()).toMatchSnapshot("Project header error");
    });

    it("should render without error when there are less than 6 users", async () => {
        const projectOverviewData: ProjectOverviewData = {
            project: {
                id: "projectIdTest",
                name: "the Test Project",
                pathWithNamespace: "/test/project",
                users: [{
                    id: "userId",
                    name: "Test user",
                    avatarUrl: "Test Avatar Url"
                }]
            }
        };

        const projectOverviewMock = {
            request: {
                query: PROJECT_OVERVIEW,
                variables: { projectId: "projectIdTest" }
            },
            result: { data: projectOverviewData }
        };

        const { asFragment } = render(
            <MockedProvider mocks={[projectOverviewMock]} addTypename={false}>
                <ProjectHeader projectId="projectIdTest" />
            </MockedProvider>
        );

        expect(screen.getByLabelText("Project overview is loading")).toBeInTheDocument();

        expect(asFragment()).toMatchSnapshot("Project Header loading");

        await waitFor(() => screen.getByLabelText("Project name: THE TEST PROJECT"))

        expect(asFragment()).toMatchSnapshot("Project Header loaded");
    });

    it("should render without error when there are more than 5 users", async () => {
        const projectOverviewData: ProjectOverviewData = {
            project: {
                id: "projectIdTest",
                name: "the Test Project",
                pathWithNamespace: "/test/project",
                users: [
                    {
                        id: "userId0",
                        name: "name0",
                        avatarUrl: "avatarUrl0"
                    },
                    {
                        id: "userId1",
                        name: "name1",
                        avatarUrl: "avatarUrl1"
                    },
                    {
                        id: "userId2",
                        name: "name2",
                        avatarUrl: "avatarUrl2"
                    },
                    {
                        id: "userId3",
                        name: "name3",
                        avatarUrl: "avatarUrl3"
                    },
                    {
                        id: "userId4",
                        name: "name4",
                        avatarUrl: "avatarUrl4"
                    },
                    {
                        id: "userId5",
                        name: "name5",
                        avatarUrl: "avatarUrl5"
                    },
                ]
            }
        };

        const projectOverviewMock = {
            request: {
                query: PROJECT_OVERVIEW,
                variables: { projectId: "projectIdTest" }
            },
            result: { data: projectOverviewData }
        };

        const { asFragment } = render(
            <MockedProvider mocks={[projectOverviewMock]} addTypename={false}>
                <ProjectHeader projectId="projectIdTest" />
            </MockedProvider>
        );

        expect(screen.getByLabelText("Project overview is loading")).toBeInTheDocument();

        await waitFor(() => screen.getByLabelText("Project name: THE TEST PROJECT"))

        fireEvent.click(screen.getByLabelText("Open project users dialog"));

        expect(asFragment()).toMatchSnapshot("Project header with user dialog open");

        expect(screen.getAllByTestId("avatar-dialog")).toHaveLength(6);
    });
})