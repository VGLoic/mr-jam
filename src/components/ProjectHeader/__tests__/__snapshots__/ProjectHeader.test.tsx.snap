// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectHeader /> should render with error: Project header error 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-item"
  >
    <h6
      aria-label="Project with id projectIdTest has not been found"
      class="MuiTypography-root MuiTypography-h6 MuiTypography-colorTextPrimary"
      data-testid="project-overview-error"
    >
      This project has not been found. Please select another one.
    </h6>
  </div>
</DocumentFragment>
`;

exports[`<ProjectHeader /> should render without error when there are less than 6 users: Project Header loaded 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-justify-xs-space-between MuiGrid-grid-xs-12"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-sm-8"
    >
      <h6
        aria-label="Project name: THE TEST PROJECT"
        class="MuiTypography-root makeStyles-projectTitle-154 MuiTypography-h6 MuiTypography-colorTextPrimary"
        data-testid="project-name"
      >
        THE TEST PROJECT
      </h6>
      <div
        class="MuiAvatarGroup-root"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circle MuiAvatarGroup-avatar makeStyles-avatar-151"
          data-testid="user-avatar"
          style="z-index: 1;"
          title="Test user"
        >
          <img
            alt="Test user"
            class="MuiAvatar-img"
            src="Test Avatar Url"
          />
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-justify-xs-flex-end MuiGrid-grid-sm-4"
    >
      <p
        class="MuiTypography-root makeStyles-fullPathLabel-155 MuiTypography-body1 MuiTypography-colorTextPrimary"
      >
        Full path:
      </p>
      <p
        aria-label="Full path: /test/project"
        class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
        data-testid="full-path"
      >
        /test/project
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ProjectHeader /> should render without error when there are less than 6 users: Project Header loading 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-item"
  >
    <div
      aria-label="Project overview is loading"
      class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
      data-testid="project-overview-loading"
      role="progressbar"
      style="width: 40px; height: 40px;"
    >
      <svg
        class="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ProjectHeader /> should render without error when there are more than 5 users: Project header with user dialog open 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-justify-xs-space-between MuiGrid-grid-xs-12"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-sm-8"
    >
      <h6
        aria-label="Project name: THE TEST PROJECT"
        class="MuiTypography-root makeStyles-projectTitle-339 MuiTypography-h6 MuiTypography-colorTextPrimary"
        data-testid="project-name"
      >
        THE TEST PROJECT
      </h6>
      <div
        class="MuiAvatarGroup-root"
      >
        <div
          class="makeStyles-avatarWrapper-337 MuiAvatarGroup-avatar"
          style="z-index: 1;"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-336"
            data-testid="user-avatar"
            title="name0"
          >
            <img
              alt="name0"
              class="MuiAvatar-img"
              src="avatarUrl0"
            />
          </div>
          <div
            class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-336"
            data-testid="user-avatar"
            title="name1"
          >
            <img
              alt="name1"
              class="MuiAvatar-img"
              src="avatarUrl1"
            />
          </div>
          <div
            class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-336"
            data-testid="user-avatar"
            title="name2"
          >
            <img
              alt="name2"
              class="MuiAvatar-img"
              src="avatarUrl2"
            />
          </div>
          <div
            class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-336"
            data-testid="user-avatar"
            title="name3"
          >
            <img
              alt="name3"
              class="MuiAvatar-img"
              src="avatarUrl3"
            />
          </div>
          <div
            class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-336"
            data-testid="user-avatar"
            title="name4"
          >
            <img
              alt="name4"
              class="MuiAvatar-img"
              src="avatarUrl4"
            />
          </div>
          <div
            aria-label="Open project users dialog"
            class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-336 makeStyles-seeMore-338 MuiAvatar-colorDefault"
            data-testid="avatar-open-dialog"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorTextPrimary"
            >
              +1
            </h6>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-justify-xs-flex-end MuiGrid-grid-sm-4"
    >
      <p
        class="MuiTypography-root makeStyles-fullPathLabel-340 MuiTypography-body1 MuiTypography-colorTextPrimary"
      >
        Full path:
      </p>
      <p
        aria-label="Full path: /test/project"
        class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
        data-testid="full-path"
      >
        /test/project
      </p>
    </div>
  </div>
</DocumentFragment>
`;
